vpr/src/draw/draw_basic.cpp

```diff
@@ -1103,6 +1103,87 @@ void draw_crit_path(ezgl::renderer* g) {
     }
 }
 
+/**
+ * @brief Draw critical path elements.
+ * 
+ * This function draws critical path elements based on the provided timing paths
+ * and indexes map. It is primarily used in server mode, where items are drawn upon request.
+ */
+void draw_crit_path_elements(const std::vector<tatum::TimingPath>& paths, const std::map<std::size_t, std::set<std::size_t>>& indexes, ezgl::renderer* g)
+{
+    t_draw_state* draw_state = get_draw_state_vars();
+    static ezgl::color contour_color{0,0,0,40};
+    const bool draw_crit_path_contour = g_vpr_ctx.server().draw_crit_path_contour();
+
+    auto drawFlylineTimingEdgeHelperFn = [](ezgl::renderer* renderer, const ezgl::color& color, ezgl::line_dash line_style, int line_width, float delay, 
+                                            const tatum::NodeId& prev_node, const tatum::NodeId& node, bool skip_draw_delays=false){
+        renderer->set_color(color);
+        renderer->set_line_dash(line_style);
+        renderer->set_line_width(line_width);
+        draw_flyline_timing_edge(tnode_draw_coord(prev_node),
+                                tnode_draw_coord(node), delay, renderer, skip_draw_delays);
+
+        renderer->set_line_dash(ezgl::line_dash::none);
+        renderer->set_line_width(0);                        
+    };
+
+    for (const auto& [pathIndex, elementIndexes]: indexes) {
+        if (pathIndex < paths.size()) {
+            const tatum::TimingPath& path = paths[pathIndex];
+
+            //Walk through the timing path drawing each edge
+            tatum::NodeId prev_node;
+            float prev_arr_time = std::numeric_limits<float>::quiet_NaN();
+            int i = 0;
```

Comment:
resolved https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/2526/commits/b9b33d8c6798cd8fa6fa273ddb2222bf2aa2eaac
Note: i is completely removed, because existed variable elementsCounter may be used instead